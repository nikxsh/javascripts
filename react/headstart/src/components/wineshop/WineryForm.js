import React, { useEffect, useState } from 'react';
import PropsTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { getWinery } from '../api/wineShopApi';
import TextInput from '../shared/TextInput';
import SelectInput from '../shared/selectInput';
import { getAllCountries } from '../api/commonApi'
import { saveWinery } from './actions/wineshopActions';

//Function component
const WineryForm = ({
	match,
	history,
	saveWinery,
	...props
}) => {
	const [winery, setWinery] = useState({ ...props.winery });
	const [saving, toggleSaving] = useState(false);
	const [countries, setCountries] = useState([]);

	useEffect(() => {
		setWinery({ ...props.winery });
		setCountries(getAllCountries());
	}, []);

	function handleChange(event) {
		event.preventDefault();
		const { name, value } = event.target;
		setWinery(prevWinery => ({
			...prevWinery,
			[name]: value
		}));
	}

	function handleSave(event) {
		event.preventDefault();
		toggleSaving(prev => !prev);
		saveWinery(winery).then(() => {
			toggleSaving(prev => !prev);
			history.push('/wineries');
		});
	}

	return <>
		<form onSubmit={handleSave}>
			<h4>{winery.id ? 'Edit' : 'Add'} Winery</h4>
			<br />
			<TextInput
				name="name"
				label="Name"
				placeholder="Winery Name"
				value={winery.name}
				onChange={handleChange}
			/>
			<TextInput
				name="region"
				label="Region"
				placeholder="Winery Region"
				value={winery.region}
				onChange={handleChange}
			/>
			<SelectInput
				name="country"
				label="Country"
				options={countries}
				value={winery.country}
				onChange={handleChange}
			/>
			<div className="form-row">
				<button type="submit" className="btn btn-primary">
					{saving ? "Saving..." : "Save"}
				</button>
				&nbsp;
			    <Link className="btn btn-warning" to={'/wineries'}>Back</Link>
			</div>
		</form>
		<div className="alert alert-light" >
		</div>
	</>
};

WineryForm.propType = {
	match: PropsTypes.object.isRequired,
	winery: PropsTypes.object.isRequired,
	history: PropsTypes.object.isRequired,
	saveWinery: PropsTypes.func.isRequired
}

export function getWineryById(wineries, id) {
	return wineries.find(winery => winery.id === id);
}

function mapStateToProps(state, ownProps) {
	const wineryId = ownProps.match.params.id;
	return {
		winery: wineryId ? getWineryById(state.wineries, wineryId) : { name: '', region: '', country: '' }
	};
}

const mapDispatchToProps = {
	saveWinery
}

export default connect(mapStateToProps, mapDispatchToProps)(WineryForm);