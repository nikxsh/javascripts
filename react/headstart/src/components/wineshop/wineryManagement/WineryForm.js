import React from 'react';

import TextInput from '../../shared/TextInput';
import SelectInput from '../../shared/selectInput';

const WineryForm = ({
	countries,
	winery,
	handleChange,
	handleSave,
	formInvalid,
	redirect,
	saving = false
}) => {

	const countryCode = countries.length > 0 ? countries.find(x => x.name === winery.country).code : 'IN';

	return <>
		<form onSubmit={handleSave}>
			<h4>{winery.id ? 'Edit' : 'Add'} Winery</h4>
			<br />
			<TextInput
				name="name"
				label="Name"
				placeholder="Winery Name"
				value={winery.name}
				onChange={handleChange}
				isRequired={true}
			/>
			<TextInput
				name="region"
				label="Region"
				placeholder="Winery Region"
				value={winery.region}
				onChange={handleChange}
				isRequired={true}
			/>
			<SelectInput
				name="country"
				label="Country"
				options={countries}
				value={countryCode}
				onChange={handleChange}
			/>
			<br />
			<div className="form-row">
				<button type="submit" className="btn btn-info mr-2" disabled={formInvalid}>
					{saving ? "Saving..." : "Save"}
				</button>
				<button type="button" className="btn btn-info" onClick={redirect}>Back</button>
			</div>
		</form>
	</>
};

export default WineryForm;