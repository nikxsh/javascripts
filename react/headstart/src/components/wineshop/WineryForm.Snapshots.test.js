import React from 'react';
import WineryForm from './WineryForm';
import renderer from 'react-test-renderer';

import { mockWineries } from '../../Test/WineryMockData';

it("Set submit button label 'Saving..' when saving is true", () => {
	const tree = renderer.create(
		<WineryForm
			countries={[]}
			winery={mockWineries[0]}
			handleChange={jest.fn()}
			handleSave={jest.fn()}
			saving={true}
		/>
	);
	expect(tree).toMatchSnapshot();
});

it("Set submit button label 'Save' when saving is false", () => {
	const tree = renderer.create(
		<WineryForm
			countries={[]}
			winery={mockWineries[1]}
			handleChange={jest.fn()}
			handleSave={jest.fn()}
			saving={false}
		/>
	);
	expect(tree).toMatchSnapshot();
});