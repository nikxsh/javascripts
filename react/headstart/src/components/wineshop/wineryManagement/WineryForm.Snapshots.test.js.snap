// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Set submit button label 'Save' when saving is false 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h4>
    Edit
     Winery
  </h4>
  <br />
  <div
    className="form-row"
  >
    <div
      className="form-group col-md-6"
    >
      <label>
        Name
      </label>
      <input
        className="form-control"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Winery Name"
        type="text"
        value="Bodegas Godeval"
      />
      <div
        className=""
      >
        
      </div>
    </div>
  </div>
  <div
    className="form-row"
  >
    <div
      className="form-group col-md-6"
    >
      <label>
        Region
      </label>
      <input
        className="form-control"
        id="region"
        name="region"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Winery Region"
        type="text"
        value="Valdeorras"
      />
      <div
        className=""
      >
        
      </div>
    </div>
  </div>
  <div
    className="form-row"
  >
    <div
      className="form-group col-md-6"
    >
      <label
        htmlFor="options"
      >
        Country
      </label>
      <select
        className="form-control"
        name="country"
        onChange={[MockFunction]}
        value="Spain"
      />
    </div>
  </div>
  <div
    className="form-row"
  >
    <button
      className="btn btn-info"
      type="submit"
    >
      Save
    </button>
  </div>
</form>
`;

exports[`Set submit button label 'Saving..' when saving is true 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h4>
    Add
     Winery
  </h4>
  <br />
  <div
    className="form-row"
  >
    <div
      className="form-group col-md-6"
    >
      <label>
        Name
      </label>
      <input
        className="form-control"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Winery Name"
        type="text"
        value="ChÃ¢teau Coutet"
      />
      <div
        className=""
      >
        
      </div>
    </div>
  </div>
  <div
    className="form-row"
  >
    <div
      className="form-group col-md-6"
    >
      <label>
        Region
      </label>
      <input
        className="form-control"
        id="region"
        name="region"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Winery Region"
        type="text"
        value="Bordeaux"
      />
      <div
        className=""
      >
        
      </div>
    </div>
  </div>
  <div
    className="form-row"
  >
    <div
      className="form-group col-md-6"
    >
      <label
        htmlFor="options"
      >
        Country
      </label>
      <select
        className="form-control"
        name="country"
        onChange={[MockFunction]}
        value="France"
      />
    </div>
  </div>
  <div
    className="form-row"
  >
    <button
      className="btn btn-info"
      type="submit"
    >
      Saving...
    </button>
  </div>
</form>
`;
